api/src/main/java/marquez/db/DbRetention.java

```diff
@@ -0,0 +1,112 @@
+                  table, retentionDays));
+      // Determine whether any rows met the retention policy.
+      if (totalNumOfRowsDeleted == 0) {
+        log.info("No rows older than '{}' days in '{}' table.", retentionDays, table);
+        continue;
+      }
+      log.info("Successfully deleted '{}' rows from '{}' table.", totalNumOfRowsDeleted, table);
+    }
+  }
+
+  /** Returns the number of {@code row}s deleted by the provided retention {@code query}. */
+  private static int executeDbRetentionQuery(
+      @NonNull Jdbi jdbi, @NonNull final String retentionQuery) throws DbRetentionException {
+    try {
+      return jdbi.withHandle(
+          handle -> {
+            final Stopwatch stopwatch = Stopwatch.createStarted();
+            int totalNumOfRowsDeleted = 0; // Keep count of rows deleted.
+            boolean hasMoreRows = true;
+            while (hasMoreRows) {
+              // Apply batch deletes (for improved performance) on rows in table with a creation
+              // time > date allowed in retention query.
+              final List<UUID> rowsDeleted =
+                  handle.createQuery(retentionQuery).mapTo(UUID.class).stream().toList();
+              final int numOfRowsDeleted = rowsDeleted.size();
+              totalNumOfRowsDeleted += numOfRowsDeleted;
+              hasMoreRows = !rowsDeleted.isEmpty();
+            }
+            // Measure the time elapsed to execute retention query, then return number of rows
+            // deleted.
+            stopwatch.stop();
+            log.debug(
+                "Deleted '{}' rows in '{}' secs using retention policy: {}",
+                totalNumOfRowsDeleted,
+                stopwatch.elapsed().toSeconds(),
+                retentionQuery);
+            return totalNumOfRowsDeleted;
+          });
+    } catch (StatementException errorOnDbRetentionQuery) {
+      log.error("Failed to apply retention query: {}", retentionQuery);
+      // Propagate throwable up the stack.
+      throw new DbRetentionException(errorOnDbRetentionQuery.getCause());
+    }
+  }
+
+  /** Build retention {@code query} for the provided {@code table} and {@code retentionDays}. */
+  private static String buildQuery(
+      @NonNull String queryTemplate, @NonNull String table, final int retentionDays) {
+    final String pk = pkInQueryFor(table);
+    final String timestamp = timestampInQueryFor(table);
+    return queryTemplate.formatted(
+        table, pk, pk, table, timestamp, retentionDays, BULK_DELETE_BATCH_SIZE, pk);
+  }
+
+  /** Returns {@code pk} for the provided {@code table}. */
+  private static String pkInQueryFor(@NonNull final String table) {
+    return LINEAGE_EVENTS.equals(table) ? Columns.RUN_UUID : Columns.ROW_UUID;
+  }
+
+  /** Returns {@code timestamp} for the provided {@code table}. */
+  private static String timestampInQueryFor(@NonNull final String table) {
+    return LINEAGE_EVENTS.equals(table) ? Columns.EVENT_TIME : Columns.CREATED_AT;
+  }
```

Comment:
I would create a nested enum here with the privateKey and timestamp fields for clarity.
```
public enum Table {
    DATASETS(Columns.ROW_UUID, Columns.CREATED_AT),
    JOBS(Columns.ROW_UUID, Columns.CREATED_AT),
    LINEAGE_EVENTS(Columns.RUN_UUID, Columns.EVENT_TIME);

    public final String privateKey;
    public final String timestamp;

    private Element(String privateKey, String timestamp) {
        this.privateKey = privateKey;
        this.timestamp = timestamp;
    }
}

Reply:
Could each of the enum contain `retentionQuery`? Being able to see the whole SQL being executed improves the readability of the code, even if it contains some repeated sql code.

Reply:
I avoided the enum all together by defining the SQL in-line for readability as suggested by @pawel-big-lebowski 👍